cdap-data-fabric/src/main/java/co/cask/cdap/data2/datafabric/dataset/DatasetServiceClient.java

```diff
@@ -312,11 +311,24 @@ private HttpResponse doRequest(HttpMethod method, String resource,
     }
   }
 
-  private HttpRequest.Builder processBuilder(HttpRequest.Builder builder) {
-    if (SecurityRequestContext.getUserId() != null) {
-      builder.addHeader(Constants.Security.Headers.USER_ID, SecurityRequestContext.getUserId());
+  private HttpRequest.Builder addUserIdHeader(HttpRequest.Builder builder) throws IOException {
+    if (!securityEnabled) {
+      return builder;
+    }
+    // If the request originated from the router and was forwarded to any service other than dataset service, before
+    // going to dataset service via dataset service client, the userId could be set in the SecurityRequestContext.
+    // e.g. deploying an app that contains a dataset
+    String userId = SecurityRequestContext.getUserId();
```

Comment:
I'm creating a PR for this - It injects an AuthenticationContext, which can tell you the user id to use depending on the context in which the call is being made. This branch has the changes - https://github.com/caskdata/cdap/compare/feature/authentication-context?expand=1. Going to create it right after this PR.